### PR TITLE
WIP: integration with prettier

### PR DESCRIPTION
**Pull request by farskid**
Following the discussion issue on https://github.com/dawnlabs/carbon/issues/443

## Description

For now it's only a naive simple formatCode using prettier's standalone package. It only uses Babylon parser.

Plan for changes:

* Smart parser switch: If the language is *auto* it should use babylon as default parser. Based on the supported list of languages built-in to the prettier and the detected language on Carbon, it should switch the parser if required.

* Showing an alert/toast in case format can't be done (prettier can be tricky sometimes)

* Adding prettier's options available in the settings dropdown

* Do the formatting in a web worker in case the code is a bit long.

@mfix22 @briandennis

### CODE REVIEW

File: lib/util.js [hunk 1]
```diff
@@ -40,3 +42,16 @@ export const fileToDataURL = blob =>
     reader.onload = e => res(e.target.result)
     reader.readAsDataURL(blob)
   })
+
+export const formatCode = code =>
+  new Promise((resolve, reject) => {
+    try {
+      const formatted = prettier.format(code, {
```
Comment (mfix22) [inline lib/util.js]: If `prettier.format` is synchronous please make `formatCode` just return the code, not a promise.

File: components/Editor.js [hunk 2]
```diff
@@ -251,15 +262,24 @@ class Editor extends React.Component {
...
-                  onClick={this.upload}
-                  title={this.state.uploading ? 'Loading...' : 'Tweet Image'}
-                  color="#57b5f9"
-                  style={{ marginRight: '8px' }}
-                />
-              )}
+              <Button
+                className="tweetButton"
+                onClick={() => this.format(this.state.code)}
+                title={this.state.formatLoading ? 'Formatting...' : 'P'}
```
Comment (mfix22) [inline components/Editor.js]: Remove `state.formatLoading` since formatting is synchronous.

File: lib/util.js (hunk 1 above)
Comment (farskid) [reply lib/util.js]: How should we handle format errors if it's sync? return Error object in the catch block and delegate error handling in the Editor?

**Comment by mfix22** [comment]
@jakedex What do you think about making the `prettier` button be the actually [Prettier logo](https://prettier.io/icon.png), or a black and white version of it.

File: components/Editor.js (hunk 2 above)
Comment (farskid) [reply components/Editor.js]: Done.

**Comment by mfix22** [comment]
screenshot 2018-08-13 14 04 46

Obviously not final but just for inspo

**Comment by farskid** [comment]
@mfix22 Should I work on the parser switch and re open this PR with that?

**Comment by farskid** [comment]
@mfix22 Not long at all. Should go into the constants and switch pasrer based on lang. I can open new pr.

**Comment by farskid** [comment]
It's more convenient to send it to another 3rd party API. I already created a prettier-server module which is doing this parser switching. I just don't know if you're interested in using it.

About the parser sizes, the babylon parser we're using right now as default is 185KB, graphql parser is 36KB, flow parser is 1.4MB, typescript parser is 2MB, ...
They're surprisingly huge!

**Comment by mfix22** [comment]
As long as we code split the custom parsers out, I’m fine with loading it
in the browser.

Which I guess will make formatCode async again 😂.

**Comment by farskid** [comment]
@mfix22 LOL, yeah I guess 😂. So I'll just work on it on a separate branch then and mention you. BTW, when is this merged branch gonna be deployed? with the next release?

**Comment by farskid** [comment]
LOL, Yeah I guess it should be async again 😂. BTW, when is branch gonna be deployed? with the next release?

**Comment by mfix22** [comment]
Yup! But I want to get a few details right before releasing. Like the code splitting to keep our bundle a little smaller.

**Pull request merged**